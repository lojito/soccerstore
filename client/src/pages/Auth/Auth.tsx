import React, { useEffect, useRef, useState, type FC } from "react";
import { styled } from "styled-components";

const Section = styled.section`
  margin: 200px auto 0;
  padding: 0 16px 16px;
  border: 1px solid #3b0062;
  border-radius: 10px;
  width: 40%;
  box-sizing: border-box;
  background-color: black;

  & .error {
    color: #fff;
    margin-top: 16px;
  }
`;

const InputWrapper = styled.div`
  margin: 20px 0;
  width: 100%;
`;

const Input = styled.input`
  display: block;
  padding: 2px 5px;
  width: 97%;
  border-radius: 3px;
  border: 1px solid #ccc;
  height: 20px;
`;

const Label = styled.label`
  display: block;
  color: #fff;
  margin-bottom: 5px;
`;

const ButtonWrapper = styled.div`
  text-align: right;
  margin-bottom: 5px;
`;

export type Account = {
  email: string;
  name?: string;
  password: string;
};

interface Props {
  onSignup?: (account: Account) => void;
  onLogin?: (account: Account) => void;
  authType: string;
  logInError?: string;
  signUpError?: string;
  isLoggedIn: boolean;
}

const Auth: FC<Props> = ({
  onSignup,
  onLogin,
  authType,
  logInError,
  signUpError,
  isLoggedIn,
}) => {
  const [account, setAccount] = useState<Account>({
    email: "",
    name: "",
    password: "",
  });

  const emailRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    !isLoggedIn && emailRef!.current!.focus();
  }, [isLoggedIn]);

  const handleSubmit = () => {
    if (authType === "signup") {
      onSignup!(account);
    } else {
      onLogin!(account);
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setAccount({
      ...account,
      [e.target.id]: e.target.value,
    });
  };

  const btnName = authType === "signup" ? "Inscribirse" : "Iniciar sesión";

  if (!isLoggedIn) {
    return (
      <Section>
        <InputWrapper>
          <Label htmlFor="email">Correo electrónico *</Label>
          <Input
            id="email"
            type="email"
            autoComplete="email"
            required
            value={account.email}
            placeholder="Correo electrónico"
            onChange={handleChange}
            ref={emailRef}
          />
        </InputWrapper>

        {authType === "signup" && (
          <InputWrapper>
            <Label htmlFor="name">Nombre *</Label>
            <Input
              id="name"
              type="text"
              required
              value={account.name}
              placeholder="Nombre"
              onChange={handleChange}
            />
          </InputWrapper>
        )}

        <InputWrapper>
          <Label htmlFor="password">Contraseña *</Label>
          <Input
            id="password"
            type="password"
            autoComplete="current-password"
            required
            value={account.password}
            placeholder="Contraseña"
            onChange={handleChange}
          />
        </InputWrapper>

        <ButtonWrapper>
          <button type="button" name={btnName} onClick={handleSubmit}>
            {btnName}
          </button>
        </ButtonWrapper>

        {signUpError && (
          <div className="error" data-testid="signUpError">
            {signUpError}
          </div>
        )}
        {logInError && (
          <div className="error" data-testid="logInError">
            {logInError}
          </div>
        )}
      </Section>
    );
  }

  return "";
};

export default Auth;
