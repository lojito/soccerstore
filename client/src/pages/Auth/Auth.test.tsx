import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Auth from "./Auth";

vi.mock("../../../hooks/auth/useAuth");

describe("Auth", () => {
  const handleSignup = vi.fn();
  const handleLogin = vi.fn();

  interface Props {
    authType?: string;
    logInError?: string;
    signUpError?: string;
    isLoggedIn?: boolean;
  }

  const TestComponent = ({
    authType = "signup",
    logInError = "",
    signUpError = "",
    isLoggedIn = false,
  }: Props) => {
    return (
      <Auth
        onSignup={handleSignup}
        onLogin={handleLogin}
        authType={authType}
        logInError={logInError}
        signUpError={signUpError}
        isLoggedIn={isLoggedIn}
      />
    );
  };

  const renderAuth = ({
    authType = "signup",
    logInError = "",
    signUpError = "",
    isLoggedIn = false,
  }: Props = {}) => {
    const utils = render(
      <TestComponent
        authType={authType}
        logInError={logInError}
        signUpError={signUpError}
        isLoggedIn={isLoggedIn}
      />
    );

    const emailLabel = screen.queryByLabelText("Correo electrónico *");
    const email = screen.queryByRole("textbox", {
      name: "Correo electrónico *",
    });

    const nameLabel = screen.queryByLabelText("Nombre *");
    const name = screen.queryByRole("textbox", {
      name: "Nombre *",
    });

    const passwordLabel = screen.queryByLabelText("Contraseña *");

    const signUpButton = screen.queryByRole("button", {
      name: "Inscribirse",
    });
    const loginButton = screen.queryByRole("button", {
      name: "Iniciar sesión",
    });

    return {
      ...utils,
      handleSignup,
      handleLogin,
      emailLabel,
      email,
      nameLabel,
      name,
      passwordLabel,
      signUpButton,
      loginButton,
    };
  };

  it("should render correctly", () => {
    const { asFragment } = renderAuth();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the signup form", () => {
    const {
      emailLabel,
      email,
      nameLabel,
      name,
      passwordLabel,
      signUpButton,
      loginButton,
    } = renderAuth();

    expect(emailLabel).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(nameLabel).toBeInTheDocument();
    expect(name).toBeInTheDocument();
    expect(passwordLabel).toBeInTheDocument();
    expect(signUpButton).toBeInTheDocument();
    expect(loginButton).not.toBeInTheDocument();
  });

  it("should render the login form", () => {
    const {
      emailLabel,
      email,
      nameLabel,
      name,
      passwordLabel,
      signUpButton,
      loginButton,
    } = renderAuth({ authType: "login" });

    expect(emailLabel).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(nameLabel).not.toBeInTheDocument();
    expect(name).not.toBeInTheDocument();
    expect(passwordLabel).toBeInTheDocument();
    expect(signUpButton).not.toBeInTheDocument();
    expect(loginButton).toBeInTheDocument();
  });

  it("should not render the login form is the user is already logged in", () => {
    const {
      emailLabel,
      email,
      nameLabel,
      name,
      passwordLabel,
      signUpButton,
      loginButton,
    } = renderAuth({ authType: "login", isLoggedIn: true });

    expect(emailLabel).not.toBeInTheDocument();
    expect(email).not.toBeInTheDocument();
    expect(nameLabel).not.toBeInTheDocument();
    expect(name).not.toBeInTheDocument();
    expect(passwordLabel).not.toBeInTheDocument();
    expect(signUpButton).not.toBeInTheDocument();
    expect(loginButton).not.toBeInTheDocument();
  });

  it("should call the auth handler when signing up the user", async () => {
    const user = userEvent.setup();
    const { signUpButton, handleSignup } = renderAuth();

    await user.click(signUpButton!);

    expect(handleSignup).toHaveBeenCalled();
  });

  it("should call the auth handler when logging in the user", async () => {
    const user = userEvent.setup();
    const { loginButton, handleLogin } = renderAuth({ authType: "login" });

    await user.click(loginButton!);

    expect(handleLogin).toHaveBeenCalled();
  });

  it("should return an error if the signing up process did fail", async () => {
    const user = userEvent.setup();
    const { signUpButton, rerender } = renderAuth();

    await user.click(signUpButton!);

    rerender(
      <TestComponent signUpError="Se produjo un error al intentar registrar al usuario." />
    );

    const errorDiv = screen.queryByTestId("signUpError");

    expect(errorDiv).toBeInTheDocument();
  });

  it("should return an error if the logging in process did fail", async () => {
    const user = userEvent.setup();
    const { loginButton, rerender } = renderAuth({ authType: "login" });

    await user.click(loginButton!);

    rerender(
      <TestComponent
        authType="login"
        logInError="Se produjo un error al intentar iniciar la sesión del usuario."
      />
    );

    const errorDiv = screen.queryByTestId("logInError");

    expect(errorDiv).toBeInTheDocument();
  });

  it("should not return an error if the logging in process did not fail", async () => {
    const user = userEvent.setup();
    const { loginButton, rerender } = renderAuth({ authType: "login" });

    await user.click(loginButton!);

    rerender(<TestComponent authType="login" />);
    const errorDiv = screen.queryByTestId("logInError");

    expect(errorDiv).not.toBeInTheDocument();
  });

  it("should allow the user to enter an email", async () => {
    const user = userEvent.setup();
    const userEmail = "user@gmail.com";
    const { email } = renderAuth({ authType: "login" });

    await user.type(email!, userEmail);

    expect(email).toHaveValue(userEmail);
  });

  it("should allow the user to enter a name", async () => {
    const user = userEvent.setup();
    const userName = "User";
    const { name } = renderAuth();

    await user.type(name!, userName);

    expect(name).toHaveValue(userName);
  });

  it("should allow the user to enter a password", async () => {
    const user = userEvent.setup();
    const userPassword = "password";
    const { passwordLabel } = renderAuth();

    await user.type(passwordLabel!, userPassword);

    expect(passwordLabel).toHaveValue(userPassword);
  });
});
