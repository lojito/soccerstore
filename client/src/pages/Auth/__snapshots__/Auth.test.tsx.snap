// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Auth > should render correctly 1`] = `
<DocumentFragment>
  <section
    class="sc-beySPh bHYugo"
  >
    <div
      class="sc-guDLey dZDvpb"
    >
      <label
        class="sc-hLQSwg faudQI"
        for="email"
      >
        Correo electr칩nico *
      </label>
      <input
        autocomplete="email"
        class="sc-dmyCSP kwjocc"
        id="email"
        placeholder="Correo electr칩nico"
        required=""
        type="email"
        value=""
      />
    </div>
    <div
      class="sc-guDLey dZDvpb"
    >
      <label
        class="sc-hLQSwg faudQI"
        for="name"
      >
        Nombre *
      </label>
      <input
        class="sc-dmyCSP kwjocc"
        id="name"
        placeholder="Nombre"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-guDLey dZDvpb"
    >
      <label
        class="sc-hLQSwg faudQI"
        for="password"
      >
        Contrase침a *
      </label>
      <input
        autocomplete="current-password"
        class="sc-dmyCSP kwjocc"
        id="password"
        placeholder="Contrase침a"
        required=""
        type="password"
        value=""
      />
    </div>
    <div
      class="sc-eDLKkx bjYlzG"
    >
      <button
        name="Inscribirse"
        type="button"
      >
        Inscribirse
      </button>
    </div>
  </section>
</DocumentFragment>
`;
